kotlin/src/main/java/org/apache/zeppelin/kotlin/KotlinInterpreter.java

```diff
@@ -0,0 +1,189 @@
+        .maxResult(maxResult)
+        .codeOnLoad("")
+        .classPath(getImportClasspath(imports))
+        .shortenTypes(shortenTypes);
+  }
+
+  public KotlinReplProperties properties() {
+    return replProperties;
+  }
+
+  @Override
+  public void open() throws InterpreterException {
+    interpreter = new KotlinRepl(replProperties);
+
+    completer.setCtx(interpreter.getKotlinContext());
+    out = new InterpreterOutputStream(logger);
+  }
+
+  @Override
+  public void close() {
+
+  }
+
+  @Override
+  public InterpreterResult interpret(String code,
+                                     InterpreterContext context) throws InterpreterException{
+    // saving job's running thread for cancelling
+    Job<?> runningJob = getRunningJob(context.getParagraphId());
+    if (runningJob != null) {
+      runningJob.info().put("CURRENT_THREAD", Thread.currentThread());
+    }
+
+    return runWithOutput(code, context.out);
+  }
+
+  @Override
+  public void cancel(InterpreterContext context) throws InterpreterException {
+    Job<?> runningJob = getRunningJob(context.getParagraphId());
+    if (runningJob != null) {
+      Map<String, Object> info = runningJob.info();
+      Object object = info.get("CURRENT_THREAD");
+      if (object instanceof Thread) {
+        try {
+          Thread t = (Thread) object;
+          t.interrupt();
+        } catch (Throwable t) {
+          logger.error("Failed to cancel script: " + t, t);
+        }
+      }
+    }
+  }
+
+  @Override
+  public FormType getFormType() throws InterpreterException {
+    return FormType.NATIVE;
+  }
+
+  @Override
+  public int getProgress(InterpreterContext context) throws InterpreterException {
+    return 0;
+  }
+
+  @Override
+  public List<InterpreterCompletion> completion(String buf, int cursor,
+      InterpreterContext interpreterContext) throws InterpreterException {
+    return completer.completion(buf, cursor, interpreterContext);
+  }
+
+  public List<KotlinVariableInfo> getVariables() {
+    return interpreter.getVariables();
+  }
+
+  public List<KotlinFunctionInfo> getFunctions() {
+    return interpreter.getFunctions();
+  }
+
+  private Job<?> getRunningJob(String paragraphId) {
+    Job foundJob = null;
+    Collection<Job> jobsRunning = getScheduler().getAllJobs();
+    for (Job job : jobsRunning) {
+      if (job.getId().equals(paragraphId)) {
+        foundJob = job;
+      }
+    }
+    return foundJob;
+  }
+
+  /**
+   * Kotlin interpreter uses System.out for printing, so it is redirected to InterpreterOutput.
+   * Note that Scala's Console class needs separate output redirection
+   */
+  private InterpreterResult runWithOutput(String code, InterpreterOutput out) {
+    this.out.setInterpreterOutput(out);
+
+    PrintStream oldOut = System.out;
+    PrintStream newOut = (out != null) ? new PrintStream(out) : null;
+    System.setOut(newOut);
+    InterpreterResult res = interpreter.eval(code);
+    System.setOut(oldOut);
```

Comment:
It's better to it in `try catch finally`

Reply:
done